// @google-cloud/common/build/src/service-object.js
"use strict";
/*!
 * Copyright 2015 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/*!
 * @module common/service-object
 */
const promisify_1 = require("@google-cloud/promisify");
const arrify = require("arrify");
const events_1 = require("events");
const extend = require("extend");
const util_1 = require("./util");
/**
 * ServiceObject is a base class, meant to be inherited from by a "service
 * object," like a BigQuery dataset or Storage bucket.
 *
 * Most of the time, these objects share common functionality; they can be
 * created or deleted, and you can get or set their metadata.
 *
 * By inheriting from this class, a service object will be extended with these
 * shared behaviors. Note that any method can be overridden when the service
 * object requires specific behavior.
 */
// tslint:disable-next-line no-any
class ServiceObject extends events_1.EventEmitter {
    /*
     * @constructor
     * @alias module:common/service-object
     *
     * @private
     *
     * @param {object} config - Configuration object.
     * @param {string} config.baseUrl - The base URL to make API requests to.
     * @param {string} config.createMethod - The method which creates this object.
     * @param {string=} config.id - The identifier of the object. For example, the
     *     name of a Storage bucket or Pub/Sub topic.
     * @param {object=} config.methods - A map of each method name that should be inherited.
     * @param {object} config.methods[].reqOpts - Default request options for this
     *     particular method. A common use case is when `setMetadata` requires a
     *     `PUT` method to override the default `PATCH`.
     * @param {object} config.parent - The parent service instance. For example, an
     *     instance of Storage if the object is Bucket.
     */
    constructor(config) {
        super();
        this.metadata = {};
        this.baseUrl = config.baseUrl;
        this.parent = config.parent; // Parent class.
        this.id = config.id; // Name or ID (e.g. dataset ID, bucket name, etc).
        this.createMethod = config.createMethod;
        this.methods = config.methods || {};
        this.interceptors = [];
        this.Promise = this.parent ? this.parent.Promise : undefined;
        this.requestModule = config.requestModule;
        if (config.methods) {
            Object.getOwnPropertyNames(ServiceObject.prototype)
                .filter(methodName => {
                return (
                // All ServiceObjects need `request`.
                // clang-format off
                !/^request/.test(methodName) &&
                    // clang-format on
                    // The ServiceObject didn't redefine the method.
                    // tslint:disable-next-line no-any
                    this[methodName] ===
                        // tslint:disable-next-line no-any
                        ServiceObject.prototype[methodName] &&
                    // This method isn't wanted.
                    !config.methods[methodName]);
            })
                .forEach(methodName => {
                // tslint:disable-next-line no-any
                this[methodName] = undefined;
            });
        }
    }
    create(optionsOrCallback, callback) {
        const self = this;
        const args = [this.id];
        if (typeof optionsOrCallback === 'function') {
            callback = optionsOrCallback;
        }
        if (typeof optionsOrCallback === 'object') {
            args.push(optionsOrCallback);
        }
        // Wrap the callback to return *this* instance of the object, not the
        // newly-created one.
        function onCreate(err, instance) {
            const args = [].slice.call(arguments);
            if (!err) {
                self.metadata = instance.metadata;
                args[1] = self; // replace the created `instance` with this one.
            }
            callback.apply(null, args);
        }
        args.push(onCreate);
        this.createMethod.apply(null, args);
    }
    delete(callback) {
        const methodConfig = (typeof this.methods.delete === 'object' && this.methods.delete) || {};
        callback = callback || util_1.util.noop;
        const reqOpts = extend({
            method: 'DELETE',
            uri: '',
        }, methodConfig.reqOpts);
        // The `request` method may have been overridden to hold any special
        // behavior. Ensure we call the original `request` method.
        this.request(reqOpts).then(res => callback(null, res), callback);
    }
    exists(callback) {
        this.get(err => {
            if (err) {
                if (err.code === 404) {
                    callback(null, false);
                }
                else {
                    callback(err);
                }
                return;
            }
            callback(null, true);
        });
    }
    get(arg0, arg1) {
        const self = this;
        let callback = arg1;
        if (typeof arg0 === 'function') {
            callback = arg0;
        }
        let config = {};
        if (typeof arg0 === 'object') {
            config = arg0;
        }
        const autoCreate = config.autoCreate && typeof this.create === 'function';
        delete config.autoCreate;
        function onCreate(err, instance, apiResponse) {
            if (err) {
                if (err.code === 409) {
                    self.get(config, callback);
                    return;
                }
                callback(err, null, apiResponse);
                return;
            }
            callback(null, instance, apiResponse);
        }
        this.getMetadata((e, metadata) => {
            const err = e;
            if (err) {
                if (err.code === 404 && autoCreate) {
                    const args = [];
                    if (Object.keys(config).length > 0) {
                        args.push(config);
                    }
                    args.push(onCreate);
                    self.create.apply(self, args);
                    return;
                }
                callback(err, null, metadata);
                return;
            }
            callback(null, self, metadata);
        });
    }
    getMetadata(callback) {
        const methodConfig = (typeof this.methods.getMetadata === 'object' &&
            this.methods.getMetadata) ||
            {};
        const reqOpts = extend({
            uri: '',
        }, methodConfig.reqOpts);
        // The `request` method may have been overridden to hold any special
        // behavior. Ensure we call the original `request` method.
        this.request(reqOpts).then(resp => {
            this.metadata = resp.body;
            callback(null, this.metadata, resp);
        }, callback);
    }
    setMetadata(metadata, callback) {
        const self = this;
        callback = callback || util_1.util.noop;
        const methodConfig = (typeof this.methods.setMetadata === 'object' &&
            this.methods.setMetadata) ||
            {};
        const reqOpts = extend(true, {
            method: 'PATCH',
            uri: '',
            json: metadata,
        }, methodConfig.reqOpts);
        // The `request` method may have been overridden to hold any special
        // behavior. Ensure we call the original `request` method.
        this.request(reqOpts).then((resp) => {
            self.metadata = resp;
            callback(null, resp);
        }, callback);
    }
    request_(reqOpts) {
        reqOpts = extend(true, {}, reqOpts);
        const isAbsoluteUrl = reqOpts.uri.indexOf('http') === 0;
        const uriComponents = [this.baseUrl, this.id || '', reqOpts.uri];
        if (isAbsoluteUrl) {
            uriComponents.splice(0, uriComponents.indexOf(reqOpts.uri));
        }
        reqOpts.uri = uriComponents
            .filter(x => x.trim()) // Limit to non-empty strings.
            .map(uriComponent => {
            const trimSlashesRegex = /^\/*|\/*$/g;
            return uriComponent.replace(trimSlashesRegex, '');
        })
            .join('/');
        const childInterceptors = arrify(reqOpts.interceptors_);
        const localInterceptors = [].slice.call(this.interceptors);
        reqOpts.interceptors_ = childInterceptors.concat(localInterceptors);
        if (reqOpts.shouldReturnStream) {
            return this.parent.requestStream(reqOpts);
        }
        return this.parent.request(reqOpts);
    }
    request(reqOpts, callback) {
        if (!callback) {
            return this.request_(reqOpts);
        }
        this.request_(reqOpts).then(res => callback(null, res.body, res), err => callback(err, err.response ? err.response.body : null, err.response));
    }
    /**
     * Make an authenticated API request.
     *
     * @private
     *
     * @param {object} reqOpts - Request options that are passed to `request`.
     * @param {string} reqOpts.uri - A URI relative to the baseUrl.
     */
    requestStream(reqOpts) {
        const opts = extend(true, reqOpts, { shouldReturnStream: true });
        return this.request_(opts);
    }
}
exports.ServiceObject = ServiceObject;
promisify_1.promisifyAll(ServiceObject, { exclude: ['requestStream', 'request', 'request_'] });
